[ Sonoma Release ] 2 Storage API tests are flaky timeouts.

whsieh opened the pull request:
#### a08ec0dd1f5593f7845b88891bae4d23f1399b88
<pre>
[ Sonoma Release ] 2 Storage API tests are flaky timeouts.
<a href="https://bugs.webkit.org/show_bug.cgi?id=277455">https://bugs.webkit.org/show_bug.cgi?id=277455</a>
<a href="https://rdar.apple.com/132941950">rdar://132941950</a>

Reviewed by Abrar Rahman Protyasha.

Adjust two API tests, IndexedDB.IndexedDBTempFileSize and WKWebView.LocalStorageProcessCrashes, so
that they&apos;re more robust. Currently, these tests both trigger actions that cause the web process to
post exactly two script messages and then use the following pattern to wait for both messages to be
received in `TestWebKitAPI`:

```
receivedScriptMessage = false;
TestWebKitAPI::Util::run(&amp;receivedScriptMessage); // &lt;--- (a)

receivedScriptMessage = false;
TestWebKitAPI::Util::run(&amp;receivedScriptMessage); // &lt;--- (b)
```

...however, it&apos;s possible for both messages to be posted right after each other in the UI process,
causing *both* messages to be received when the runloop turns in step (a), marked above. This
results in a timeout, as (b) will never return since the second message has already been received.

To avoid this flaky timeout, we can simply refactor these two tests so that we wait for the second
message to be received in the UI process, keep track of all messages received in an array, and then
check the contents of the array instead of the last received message.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBTempFileSize.mm:
(-[IndexedDBFileSizeMessageHandler userContentController:didReceiveScriptMessage:]):
(TEST(IndexedDB, IndexedDBTempFileSize)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(-[LocalStorageMessageHandler userContentController:didReceiveScriptMessage:]):
(TEST(WKWebView, LocalStorageProcessCrashes)):

Canonical link: <a href="https://commits.webkit.org/281707@main">https://commits.webkit.org/281707@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf97e7b0ce2679785a8e849b6902e270b164f3b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60707 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64638 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11254 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62837 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47742 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11529 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49085 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7803 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62741 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37306 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52579 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29914 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33995 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9813 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10167 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55841 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10111 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66367 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4651 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56453 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4672 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52555 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56634 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3847 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9133 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35871 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36953 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38046 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36698 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
```diff
@@ -105,13 +106,12 @@ - (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigat
-    TestWebKitAPI::Util::run(&receivedScriptMessage);
-    EXPECT_WK_STREQ(@"session:storage", [lastScriptMessage body]);
+    EXPECT_EQ([receivedMessages count], 2U);
+    EXPECT_WK_STREQ(@"local:storage", [receivedMessages firstObject]);
```
Does the content of the other item in this array not matter?

whsieh replied:
Ah, so we do care about the first value, but the `waitForConditionWithLogging` above is sufficient to verify that the second value is `session:storage`

Thanks for the review!

whsieh committed:
[ Sonoma Release ] 2 Storage API tests are flaky timeouts.
https://bugs.webkit.org/show_bug.cgi?id=277455
rdar://132941950

Reviewed by Abrar Rahman Protyasha.

Adjust two API tests, IndexedDB.IndexedDBTempFileSize and WKWebView.LocalStorageProcessCrashes, so
that they're more robust. Currently, these tests both trigger actions that cause the web process to
post exactly two script messages and then use the following pattern to wait for both messages to be
received in `TestWebKitAPI`:

```
receivedScriptMessage = false;
TestWebKitAPI::Util::run(&receivedScriptMessage); // <--- (a)

receivedScriptMessage = false;
TestWebKitAPI::Util::run(&receivedScriptMessage); // <--- (b)
```

...however, it's possible for both messages to be posted right after each other in the UI process,
causing *both* messages to be received when the runloop turns in step (a), marked above. This
results in a timeout, as (b) will never return since the second message has already been received.

To avoid this flaky timeout, we can simply refactor these two tests so that we wait for the second
message to be received in the UI process, keep track of all messages received in an array, and then
check the contents of the array instead of the last received message.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBTempFileSize.mm:
(-[IndexedDBFileSizeMessageHandler userContentController:didReceiveScriptMessage:]):
(TEST(IndexedDB, IndexedDBTempFileSize)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(-[LocalStorageMessageHandler userContentController:didReceiveScriptMessage:]):
(TEST(WKWebView, LocalStorageProcessCrashes)):

Canonical link: https://commits.webkit.org/281707@main
